### ISSUE TITLE
ServerOptions should notify user if changing settings while the server is running

### ISSUE THREAD

**Issue by dyfer** [supercollider/supercollider#4897]
## Motivation

`ServerOptions` (e.g. `Server.default.options`) need to be set before booting the server. It has bitten me many times when debugging other people's code when number of input/output channels was set after the server booted, which led to mysterious behavior/bugs. I've been thinking about this many times over the years; recently #4895 brought this to the surface again.

## Description of Proposed Feature

I think it would be good to notify the user if one tries to set the server options after booting the server. Some considerations:
- methods affected  - I can really see the problem only in setting number of input/output channels. Should the notification be implemented for
  - only numInputBusChannels/numOutputBusChannels OR
  - all `ServerOptions` values?
- notification behavior in case user tries to change the settings after booting the server
   - Posting a `Warning` only (and maintaining current behavior) OR
   - Throwing an `Error` and not setting the parameters (I believe this would be an API change?)

## Plan for Implementation

The implementation would add a method/variable to `ServerOptions` that the `Server` class would set when booting/quitting. Its value would be checked agains in the parameter changing methods. I'm happy to implement that once we agree on exact features.

Please let me know what you think!

**Comment by brianlheim** [comment]
i disagree with the way you want to solve this -- firstly, in a good OO design the options object should know nothing about any server(s) using it, and secondly there are really legitimate reasons for changing server options while the server holding it is attached to a running process, and we don't want to penalize that by printing spurious warnings -- but i do think it should be easier to see the options the server was booted with. what do you think about the Server class providing a method that prints out the command used to boot the server process? this would be `nil` when there is no active server process. in fact, why not have the server class print that when it boots the process? that would give a chance to notice anything amiss.

**Comment by dyfer** [comment]
I understand the issue with OO design, but in real life use case it's a considerable problem:
```supercollider
s.waitForBoot({
  // do stuff
  //...
  s.options.numOutputBusChannels = 16;
  // do more stuff
  // things might behave differently here...
})
```
Particularly, you would often depend on `s.options.numOutputBusChannels` to set synths's input buses (to read from HW input) midway in the project. If you change that option while the server is running, you are getting the wrong bus number.

So IIUC, the "proper" solution would be to set a "number in/out channels" parameter for the Server object. I just want to point out that this is potentially a big API change - majority of projects I deal with are using `s.options.numOutputBusChannels` (which I understand would still be there, but it's just not working reliably in the case I described). Also, getting the information through printing is not solving anything. I think it's an issue of accessing the bus numbers reliably for addressing input buses programmatically.

Also I'm curious, what are legitimate reasons for changing server options while the server is running? I can only imagine this to be the case as a preparation step before a reboot.

**Comment by brianlheim** [comment]
> I can only imagine this to be the case as a preparation step before a reboot.

yes, that's the case i was thinking of.

> If you change that option while the server is running, you are getting the wrong bus number.

i'm really sorry, but i don't understand what exactly you're talking about now.. isn't that a programming error? when is that ever expected to work?

**Comment by dyfer** [comment]
> i'm really sorry, but i don't understand what exactly you're talking about now.. isn't that a programming error? when is that ever expected to work?

I was referring to changing changing options' numIn/OutputBusChannels while the server is running. After you do that, your attempt to get sound from hardware input, which AFAIK, can only be done by querying ServerOptions, will likely not correlate with the hardware input.

IMO it _is_ programming error when you change num IO channels in server options while the server is running, but there's no way to tell (i.e. no warning or error) so you won't know immediately, but you possibly won't be getting sound from hardware input in your synth.

Please let me know if this makes sense. It is crucial to me that it is clear what the use-case issue is.

**Comment by dyfer** [comment]
AfterI updated this issue after private conversation with @brianlheim
